File: ios/App/App/FloodDetector.swift
import Foundation

// MARK: - FloodDetector

/// Reply-flood detection for notification events — a Swift port of the web
/// client's `src/lib/replyFlood.ts` (and Android's `FloodDetector.java`). See
/// that file for the full rationale; the rules and constants here mirror it
/// exactly so a spam wall the web UI folds out of a thread is the same wall
/// this suppresses from iOS notifications.
///
/// A DISPLAY heuristic, not a moderation boundary. It never blocks a write or
/// feeds a report — it only decides which events belong to a visual flood and
/// should be suppressed. Two invariants the caller relies on:
///
///  1. Never flag the reading user's own event (`selfPubkey`).
///  2. Never flag an event from someone the reader follows (`follows`).
///
/// Both exemptions are applied LAST, so they override every rule at once.
///
/// What survives from the port:
///  - ECHO: one near-duplicate template posted across `echoMinAuthors`+ pubkeys.
///  - DENSITY: one pubkey repeating the same template `densityMin`+ times.
///  - A CONTAINMENT sweep that only WIDENS an already-confirmed flood.
///
/// The caller feeds a batch (a poll cycle's worth of events) so the crowd
/// needed by ECHO/DENSITY is visible. Pure and batch-local: it reads only the
/// events passed in.
enum FloodDetector {

    // MARK: - Tunables (mirror replyFlood.ts)

    /// Distinct pubkeys one template must span before it reads as a campaign.
    static let echoMinAuthors = 3
    /// Copies of one template (across any authors) before the echo rule fires.
    static let echoMinCopies = 3
    /// Copies of one template from a SINGLE pubkey before the density rule fires.
    static let densityMin = 4
    /// Words a template needs before either rule will judge it.
    static let minWords = 4
    /// The lower word bar for a template that carries a link.
    static let minWordsLinked = 3
    /// Token overlap (Jaccard) at which two templates are treated as one campaign.
    static let similarityThreshold = 0.6
    /// Overlap coefficient at which a still-visible template is swept into a confirmed campaign.
    static let containmentThreshold = 0.8
    /// Words the SMALLER of two templates needs before the containment sweep will judge them.
    static let containmentMinWords = 6
    /// Document frequency above which a word is skipped when generating merge candidates.
    static let dfCap = 64
    /// Most candidate templates one template is compared against.
    static let candidateCap = 48

    // MARK: - Regex (compiled once)

    private static let urlRun = try! NSRegularExpression(pattern: "https?://\\S+")
    private static let trailingNonce = try! NSRegularExpression(pattern: "([>!])\\s*[a-z0-9]{4,9}$")
    private static let digitToken = try! NSRegularExpression(pattern: "[\\p{L}\\p{N}]*\\p{N}[\\p{L}\\p{N}]*")
    /// Three-plus of one letter: elongation (nooo), laughter (kkkk), mash (ggggg).
    private static let letterRun = try! NSRegularExpression(pattern: "(\\p{L})\\1{2,}")
    private static let invisible = try! NSRegularExpression(pattern: "[\\x{200b}-\\x{200f}\\x{2060}\\x{feff}]")
    private static let whitespace = try! NSRegularExpression(pattern: "\\s+")
    /// Words, in any script. Emoji and punctuation are deliberately not words.
    private static let word = try! NSRegularExpression(pattern: "[\\p{L}][\\p{L}\\p{N}_]*")

    // MARK: - Input

    /// One event for the detector: just the fields the rules read.
    struct Event {
        let id: String
        let pubkey: String
        let content: String
    }

    // MARK: - Internal types

    /// One template and every event that shares it.
    private final class Bucket {
        let words: Set<String>
        let wordCount: Int
        let linked: Bool
        var ids: [String] = []
        var authors: Set<String> = []

        init(words: Set<String>, wordCount: Int, linked: Bool) {
            self.words = words
            self.wordCount = wordCount
            self.linked = linked
        }
    }

    private final class Campaign {
        var ids: [String] = []
        var authors: Set<String> = []
        var eligible = false
    }

    // MARK: - Regex helpers

    private static func replaceAll(_ regex: NSRegularExpression, in s: String, with template: String) -> String {
        let range = NSRange(s.startIndex..<s.endIndex, in: s)
        return regex.stringByReplacingMatches(in: s, range: range, withTemplate: template)
    }

    /// A template fingerprint: what two copies of one broadcast share once the
    /// parts that vary per copy (URLs, digits, a trailing nonce, held-down keys)
    /// are collapsed.
    static func shapeKey(_ content: String) -> String {
        var s = content.lowercased()
        s = replaceAll(invisible, in: s, with: "")
        s = replaceAll(whitespace, in: s, with: " ").trimmingCharacters(in: .whitespaces)
        s = replaceAll(urlRun, in: s, with: "@")
        s = replaceAll(trailingNonce, in: s, with: "$1#")
        s = replaceAll(digitToken, in: s, with: "#")
        // A letter run is one keypress held down: `ggg` and `gggggg` are the same
        // message, and `noooo` is `no`. Two is a word (`gg`), three is a run.
        s = replaceAll(letterRun, in: s, with: "$1")
        s = replaceAll(whitespace, in: s, with: " ").trimmingCharacters(in: .whitespaces)
        return s
    }

    /// The words of a template, for the similarity measure and the length gate.
    static func normalizeTokens(_ shape: String) -> [String] {
        let range = NSRange(shape.startIndex..<shape.endIndex, in: shape)
        var words: [String] = []
        word.enumerateMatches(in: shape, range: range) { match, _, _ in
            if let match, let r = Range(match.range, in: shape) {
                words.append(String(shape[r]))
            }
        }
        return words
    }

    private static func eligible(_ wordCount: Int, _ linked: Bool) -> Bool {
        wordCount >= (linked ? minWordsLinked : minWords)
    }

    /// Jaccard overlap of two token sets.
    private static func similarity(_ a: Set<String>, _ b: Set<String>) -> Double {
        var shared = 0
        for w in a where b.contains(w) { shared += 1 }
        return Double(shared) / Double(a.count + b.count - shared)
    }

    /// Overlap coefficient: shared words over the SMALLER set.
    private static func containment(_ a: Set<String>, _ b: Set<String>) -> Double {
        let small = a.count <= b.count ? a : b
        let big = a.count <= b.count ? b : a
        if small.isEmpty { return 0 }
        var shared = 0
        for w in small where big.contains(w) { shared += 1 }
        return Double(shared) / Double(small.count)
    }

    // MARK: - Union-find

    private static func find(_ parent: inout [Int], _ i: Int) -> Int {
        var i = i
        while parent[i] != i {
            parent[i] = parent[parent[i]]
            i = parent[i]
        }
        return i
    }

    private static func union(_ parent: inout [Int], _ i: Int, _ j: Int) {
        let ri = find(&parent, i)
        let rj = find(&parent, j)
        if ri != rj { parent[max(ri, rj)] = min(ri, rj) }
    }

    // MARK: - Detection

    /// The ids of events belonging to a visual flood. `selfPubkey` and
    /// `follows` may be nil / empty.
    static func floodIds(_ replies: [Event], selfPubkey: String?, follows: Set<String>?) -> Set<String> {
        var flagged = Set<String>()
        if replies.count < echoMinCopies { return flagged }

        // Bucket by exact normalized shape. A reply with no words joins no bucket.
        var byShape = [String: Bucket]()
        var shapeOrder: [String] = []
        for ev in replies {
            let shape = shapeKey(ev.content)
            if shape.isEmpty { continue }
            let words = normalizeTokens(shape)
            if words.isEmpty { continue }
            let bucket: Bucket
            if let existing = byShape[shape] {
                bucket = existing
            } else {
                bucket = Bucket(words: Set(words), wordCount: words.count, linked: shape.contains("@"))
                byShape[shape] = bucket
                shapeOrder.append(shape)
            }
            bucket.ids.append(ev.id)
            bucket.authors.insert(ev.pubkey)
        }

        let buckets = shapeOrder.map { byShape[$0]! }
        let n = buckets.count

        // Merge near-duplicate templates into campaigns via union-find.
        var parent = Array(0..<n)

        var elig: [Int] = []
        for i in 0..<n where eligible(buckets[i].wordCount, buckets[i].linked) {
            elig.append(i)
        }

        if elig.count >= 2 {
            // Document frequency, to drop words too common to narrow anything.
            var df = [String: Int]()
            for i in elig {
                for w in buckets[i].words { df[w, default: 0] += 1 }
            }

            // Inverted word→template index: only templates that share a word
            // are merge candidates.
            var index = [String: [Int]]()
            for i in elig {
                for w in buckets[i].words {
                    if (df[w] ?? 0) > dfCap { continue }
                    index[w, default: []].append(i)
                }
            }

            var count = Array(repeating: 0, count: n)
            var touched: [Int] = []
            for i in elig {
                touched.removeAll(keepingCapacity: true)
                for w in buckets[i].words {
                    guard let list = index[w] else { continue }
                    for j in list {
                        if j >= i { continue } // each pair once
                        if count[j] == 0 { touched.append(j) }
                        count[j] += 1
                    }
                }
                // Above the cap, compare the candidates sharing the MOST words first.
                if touched.count > candidateCap {
                    touched.sort { count[$0] > count[$1] }
                }
                let limit = min(touched.count, candidateCap)
                for k in 0..<limit {
                    let j = touched[k]
                    if find(&parent, i) != find(&parent, j)
                        && similarity(buckets[i].words, buckets[j].words) >= similarityThreshold {
                        union(&parent, i, j)
                    }
                }
                for j in touched { count[j] = 0 }
            }
        }

        // Gather campaigns (union-find roots), merging ids and author sets.
        var campaigns = [Int: Campaign]()
        for b in 0..<n {
            let root = find(&parent, b)
            let c = campaigns[root] ?? Campaign()
            c.ids.append(contentsOf: buckets[b].ids)
            c.authors.formUnion(buckets[b].authors)
            if eligible(buckets[b].wordCount, buckets[b].linked) { c.eligible = true }
            campaigns[root] = c
        }

        for c in campaigns.values {
            if !c.eligible { continue }
            let echo = c.authors.count >= echoMinAuthors && c.ids.count >= echoMinCopies
            let density = c.authors.count == 1 && c.ids.count >= densityMin
            if echo || density { flagged.formUnion(c.ids) }
        }

        // Containment sweep: only WIDENS an already-confirmed flood, never forms one.
        if !flagged.isEmpty {
            var flaggedWords: [Set<String>] = []
            var unflagged: [Int] = []
            for b in 0..<n {
                if buckets[b].ids.isEmpty { continue }
                if flagged.contains(buckets[b].ids[0]) {
                    flaggedWords.append(buckets[b].words)
                } else if buckets[b].wordCount >= containmentMinWords {
                    unflagged.append(b)
                }
            }
            for b in unflagged {
                for fw in flaggedWords {
                    if fw.count < containmentMinWords { continue }
                    if containment(buckets[b].words, fw) >= containmentThreshold {
                        flagged.formUnion(buckets[b].ids)
                        break
                    }
                }
            }
        }

        // Trust exemption, applied LAST so it overrides every rule.
        if !flagged.isEmpty && (selfPubkey != nil || !(follows?.isEmpty ?? true)) {
            for ev in replies {
                if ev.pubkey == selfPubkey || (follows?.contains(ev.pubkey) ?? false) {
                    flagged.remove(ev.id)
                }
            }
        }

        return flagged
    }
}

File: android/app/src/main/java/pub/ditto/app/FloodDetector.java
package pub.ditto.app;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Reply-flood detection for notification events — a Java port of the web
 * client's {@code src/lib/replyFlood.ts}. See that file for the full rationale;
 * the rules and constants here mirror it exactly so a spam wall the web UI folds
 * out of a thread is the same wall this suppresses from Android notifications.
 *
 * A DISPLAY heuristic, not a moderation boundary. It never blocks a write or
 * feeds a report — it only decides which notification events belong to a visual
 * flood and should be suppressed. The two invariants the caller relies on:
 *
 *  1. Never flag the reading user's own event ({@code self}).
 *  2. Never flag an event from someone the reader follows ({@code follows}).
 *
 * Both exemptions are applied LAST, so they override every rule at once.
 *
 * What survives from the port:
 *  - ECHO: one near-duplicate template posted across {@link #ECHO_MIN_AUTHORS}+
 *    distinct pubkeys.
 *  - DENSITY: one pubkey repeating the same template {@link #DENSITY_MIN}+ times.
 *  - SALAD: a crowd of throwaway pubkeys each posting a uniquely-shaped word
 *    salad drawn from one shared vocabulary pool — the campaign that defeats
 *    shape clustering on purpose. Keys off the crowd (many authors, one shared
 *    word pool), not shape.
 *  - A CONTAINMENT sweep that only WIDENS an already-confirmed flood.
 *
 * Because notifications arrive one at a time on the live socket, the caller
 * feeds a rolling window of recent events (not just the current one) so the
 * crowd needed by ECHO/DENSITY is visible. Pure and batch-local: it reads only
 * the events passed in.
 */
public final class FloodDetector {

    private FloodDetector() {}

    /** Distinct pubkeys one template must span before it reads as a campaign. */
    public static final int ECHO_MIN_AUTHORS = 3;
    /** Copies of one template (across any authors) before the echo rule fires. */
    public static final int ECHO_MIN_COPIES = 3;
    /** Copies of one template from a SINGLE pubkey before the density rule fires. */
    public static final int DENSITY_MIN = 4;
    /** Words a template needs before either rule will judge it. */
    public static final int MIN_WORDS = 4;
    /** The lower word bar for a template that carries a link. */
    public static final int MIN_WORDS_LINKED = 3;
    /** Token overlap (Jaccard) at which two templates are treated as one campaign. */
    public static final double SIMILARITY = 0.6;
    /** Overlap coefficient at which a still-visible template is swept into a confirmed campaign. */
    public static final double CONTAINMENT = 0.8;
    /** Words the SMALLER of two templates needs before the containment sweep will judge them. */
    public static final int CONTAINMENT_MIN_WORDS = 6;
    /** Distinct authors a shared word pool must span before SALAD reads as a campaign. */
    public static final int SALAD_MIN_AUTHORS = 5;
    /** Distinct words a reply needs before SALAD will judge it. */
    public static final int SALAD_MIN_WORDS = 6;
    /** Document frequency (over the distinct-author corpus) at/above which a word is "pool". */
    public static final int SALAD_POOL_DF = 3;
    /** Fraction of a reply's distinct words that must be pool words to read as salad. */
    public static final double SALAD_POOL_FRACTION = 0.6;
    /** Document frequency above which a word is skipped when generating merge candidates. */
    private static final int DF_CAP = 64;
    /** Most candidate templates one template is compared against. */
    private static final int CANDIDATE_CAP = 48;

    private static final Pattern URL_RUN = Pattern.compile("https?://\\S+");
    private static final Pattern TRAILING_NONCE = Pattern.compile("([>!])\\s*[a-z0-9]{4,9}$");
    // Note: `\p{L}`/`\p{N}` are already Unicode general-category properties in
    // Java, so no UNICODE_CHARACTER_CLASS flag is needed (and Android's regex
    // engine throws IllegalArgumentException if that flag is passed).
    private static final Pattern DIGIT_TOKEN =
            Pattern.compile("[\\p{L}\\p{N}]*\\p{N}[\\p{L}\\p{N}]*");
    /** Three-plus of one letter: elongation (nooo), laughter (kkkk), mash (ggggg). */
    private static final Pattern LETTER_RUN =
            Pattern.compile("(\\p{L})\\1{2,}");
    private static final Pattern INVISIBLE =
            Pattern.compile("[\\u200b-\\u200f\\u2060\\ufeff]");
    private static final Pattern WHITESPACE = Pattern.compile("\\s+");
    /** Words, in any script. Emoji and punctuation are deliberately not words. */
    private static final Pattern WORD =
            Pattern.compile("[\\p{L}][\\p{L}\\p{N}_]*");

    /** One event for the detector: just the fields the rules read. */
    public static final class Event {
        public final String id;
        public final String pubkey;
        public final String content;

        public Event(String id, String pubkey, String content) {
            this.id = id;
            this.pubkey = pubkey;
            this.content = content;
        }
    }

    /**
     * A template fingerprint: what two copies of one broadcast share once the
     * parts that vary per copy (URLs, digits, a trailing nonce, held-down keys)
     * are collapsed.
     */
    public static String shapeKey(String content) {
        String s = content.toLowerCase(Locale.ROOT);
        s = INVISIBLE.matcher(s).replaceAll("");
        s = WHITESPACE.matcher(s).replaceAll(" ").trim();
        s = URL_RUN.matcher(s).replaceAll("@");
        s = TRAILING_NONCE.matcher(s).replaceAll("$1#");
        s = DIGIT_TOKEN.matcher(s).replaceAll("#");
        // A letter run is one keypress held down: `ggg` and `gggggg` are the same
        // message, and `noooo` is `no`. Two is a word (`gg`), three is a run.
        s = LETTER_RUN.matcher(s).replaceAll("$1");
        s = WHITESPACE.matcher(s).replaceAll(" ").trim();
        return s;
    }

    /** The words of a template, for the similarity measure and the length gate. */
    public static List<String> normalizeTokens(String shape) {
        List<String> words = new ArrayList<>();
        Matcher m = WORD.matcher(shape);
        while (m.find()) words.add(m.group());
        return words;
    }

    /** One template and every event that shares it. */
    private static final class Bucket {
        final Set<String> words;
        final int wordCount;
        final boolean linked;
        final List<String> ids = new ArrayList<>();
        final Set<String> authors = new HashSet<>();

        Bucket(Set<String> words, int wordCount, boolean linked) {
            this.words = words;
            this.wordCount = wordCount;
            this.linked = linked;
        }
    }

    private static boolean eligible(int wordCount, boolean linked) {
        return wordCount >= (linked ? MIN_WORDS_LINKED : MIN_WORDS);
    }

    /** Jaccard overlap of two token sets. */
    private static double similarity(Set<String> a, Set<String> b) {
        int shared = 0;
        for (String w : a) if (b.contains(w)) shared++;
        return (double) shared / (a.size() + b.size() - shared);
    }

    /** Overlap coefficient: shared words over the SMALLER set. */
    private static double containment(Set<String> a, Set<String> b) {
        Set<String> small = a.size() <= b.size() ? a : b;
        Set<String> big = a.size() <= b.size() ? b : a;
        if (small.isEmpty()) return 0;
        int shared = 0;
        for (String w : small) if (big.contains(w)) shared++;
        return (double) shared / small.size();
    }

    /**
     * The ids of events belonging to a visual flood. {@code self} and
     * {@code follows} may be null.
     */
    public static Set<String> floodIds(List<Event> replies, String self, Set<String> follows) {
        Set<String> flagged = new HashSet<>();
        if (replies.size() < ECHO_MIN_COPIES) return flagged;

        // Bucket by exact normalized shape. A reply with no words joins no bucket.
        Map<String, Bucket> byShape = new LinkedHashMap<>();
        for (Event ev : replies) {
            String shape = shapeKey(ev.content);
            if (shape.isEmpty()) continue;
            List<String> words = normalizeTokens(shape);
            if (words.isEmpty()) continue;
            Bucket bucket = byShape.get(shape);
            if (bucket == null) {
                bucket = new Bucket(new HashSet<>(words), words.size(), shape.indexOf('@') >= 0);
                byShape.put(shape, bucket);
            }
            bucket.ids.add(ev.id);
            bucket.authors.add(ev.pubkey);
        }

        List<Bucket> buckets = new ArrayList<>(byShape.values());
        int n = buckets.size();

        // Merge near-duplicate templates into campaigns via union-find.
        int[] parent = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;

        List<Integer> elig = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (eligible(buckets.get(i).wordCount, buckets.get(i).linked)) elig.add(i);
        }

        if (elig.size() >= 2) {
            // Document frequency, to drop words too common to narrow anything.
            Map<String, Integer> df = new HashMap<>();
            for (int i : elig) {
                for (String w : buckets.get(i).words) {
                    df.merge(w, 1, Integer::sum);
                }
            }

            // Inverted word→template index: only templates that share a word
            // are merge candidates.
            Map<String, List<Integer>> index = new HashMap<>();
            for (int i : elig) {
                for (String w : buckets.get(i).words) {
                    if (df.getOrDefault(w, 0) > DF_CAP) continue;
                    index.computeIfAbsent(w, k -> new ArrayList<>()).add(i);
                }
            }

            int[] count = new int[n];
            List<Integer> touched = new ArrayList<>();
            for (int i : elig) {
                touched.clear();
                for (String w : buckets.get(i).words) {
                    List<Integer> list = index.get(w);
                    if (list == null) continue;
                    for (int j : list) {
                        if (j >= i) continue; // each pair once
                        if (count[j] == 0) touched.add(j);
                        count[j]++;
                    }
                }
                // Above the cap, compare the candidates sharing the MOST words first.
                if (touched.size() > CANDIDATE_CAP) {
                    touched.sort((a, b) -> count[b] - count[a]);
                }
                int limit = Math.min(touched.size(), CANDIDATE_CAP);
                for (int k = 0; k < limit; k++) {
                    int j = touched.get(k);
                    if (find(parent, i) != find(parent, j)
                            && similarity(buckets.get(i).words, buckets.get(j).words) >= SIMILARITY) {
                        union(parent, i, j);
                    }
                }
                for (int j : touched) count[j] = 0;
            }
        }

        // Gather campaigns (union-find roots), merging ids and author sets.
        Map<Integer, Campaign> campaigns = new HashMap<>();
        for (int b = 0; b < n; b++) {
            int root = find(parent, b);
            Campaign c = campaigns.get(root);
            if (c == null) {
                c = new Campaign();
                campaigns.put(root, c);
            }
            c.ids.addAll(buckets.get(b).ids);
            c.authors.addAll(buckets.get(b).authors);
            if (eligible(buckets.get(b).wordCount, buckets.get(b).linked)) c.eligible = true;
        }

        for (Campaign c : campaigns.values()) {
            if (!c.eligible) continue;
            boolean echo = c.authors.size() >= ECHO_MIN_AUTHORS && c.ids.size() >= ECHO_MIN_COPIES;
            boolean density = c.authors.size() == 1 && c.ids.size() >= DENSITY_MIN;
            if (echo || density) flagged.addAll(c.ids);
        }

        // SALAD: a crowd of throwaway pubkeys each posting a uniquely-shaped word
        // salad drawn from one shared vocabulary pool. Shape-based ECHO/DENSITY
        // are blind to it, so this reads the crowd instead: build a corpus of
        // substantial replies (one per pubkey), take the pool to be the words
        // recurring across SALAD_POOL_DF+ distinct authors, and flag replies
        // mostly built from that pool — but only when they span a big enough crowd.
        Map<String, Set<String>> perAuthor = new HashMap<>();
        for (Event ev : replies) {
            if (perAuthor.containsKey(ev.pubkey)) continue;
            List<String> words = normalizeTokens(shapeKey(ev.content));
            if (words.size() < SALAD_MIN_WORDS) continue;
            perAuthor.put(ev.pubkey, new HashSet<>(words));
        }
        if (perAuthor.size() >= SALAD_MIN_AUTHORS) {
            Map<String, Integer> poolDf = new HashMap<>();
            for (Set<String> set : perAuthor.values()) {
                for (String w : set) poolDf.merge(w, 1, Integer::sum);
            }
            List<String> memberIds = new ArrayList<>();
            Set<String> memberAuthors = new HashSet<>();
            for (Event ev : replies) {
                List<String> words = normalizeTokens(shapeKey(ev.content));
                if (words.size() < SALAD_MIN_WORDS) continue;
                Set<String> distinct = new HashSet<>(words);
                int pool = 0;
                for (String w : distinct) {
                    if (poolDf.getOrDefault(w, 0) >= SALAD_POOL_DF) pool++;
                }
                if ((double) pool / distinct.size() >= SALAD_POOL_FRACTION) {
                    memberIds.add(ev.id);
                    memberAuthors.add(ev.pubkey);
                }
            }
            if (memberAuthors.size() >= SALAD_MIN_AUTHORS) {
                flagged.addAll(memberIds);
            }
        }

        // Containment sweep: only WIDENS an already-confirmed flood, never forms one.
        if (!flagged.isEmpty()) {
            List<Set<String>> flaggedWords = new ArrayList<>();
            List<Integer> unflagged = new ArrayList<>();
            for (int b = 0; b < n; b++) {
                if (buckets.get(b).ids.isEmpty()) continue;
                if (flagged.contains(buckets.get(b).ids.get(0))) {
                    flaggedWords.add(buckets.get(b).words);
                } else if (buckets.get(b).wordCount >= CONTAINMENT_MIN_WORDS) {
                    unflagged.add(b);
                }
            }
            for (int b : unflagged) {
                for (Set<String> fw : flaggedWords) {
                    if (fw.size() < CONTAINMENT_MIN_WORDS) continue;
                    if (containment(buckets.get(b).words, fw) >= CONTAINMENT) {
                        flagged.addAll(buckets.get(b).ids);
                        break;
                    }
                }
            }
        }

        // Trust exemption, applied LAST so it overrides every rule.
        if (!flagged.isEmpty() && (self != null || (follows != null && !follows.isEmpty()))) {
            for (Event ev : replies) {
                if (ev.pubkey.equals(self) || (follows != null && follows.contains(ev.pubkey))) {
                    flagged.remove(ev.id);
                }
            }
        }

        return flagged;
    }

    private static final class Campaign {
        final List<String> ids = new ArrayList<>();
        final Set<String> authors = new HashSet<>();
        boolean eligible = false;
    }

    private static int find(int[] parent, int i) {
        while (parent[i] != i) {
            parent[i] = parent[parent[i]];
            i = parent[i];
        }
        return i;
    }

    private static void union(int[] parent, int i, int j) {
        int ri = find(parent, i);
        int rj = find(parent, j);
        if (ri != rj) parent[Math.max(ri, rj)] = Math.min(ri, rj);
    }
}
